### PR TITLE
RES: Fix handling expansion of macros inside functions

### LINKED ISSUE #9720: Plugin thinks type is imported when it is not
## Environment

* **IntelliJ Rust plugin version:** 0.4.182.4996-222
* **Rust toolchain version:** 1.62.0-beta.7 (747075d9c 2022-06-26) x86_64-unknown-linux-gnu
* **IDE name and version:** CLion 2022.2.4 (CL-222.4345.21)
* **Operating system:** Linux 5.15.0-52-generic
* **Macro expansion:** enabled
* **Additional experimental features:** org.rust.macros.proc

## Problem description
The IDE seems to think sometimes that a type is imported (thus showing up white, rather than red), hitting Ctrl+space does not do anything. The type is not imported though (and rustc confirms)

## Steps to reproduce
Honestly, I dont know for sure. It usually works, except for one type in my project. (HttpError, an enum, in my case)

I unfortunately cannot include the codebase publicly, though I can share it privately with this project's maintainers, if need be.

### PR DESCRIPTION
**Pull request by dima74**
Fixes #9720

changelog: Fix handling expansion of macros inside functions

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/macros/MacroExpansionSharedCache.kt [hunk 1]
```diff
@@ -350,6 +350,12 @@ class ExpansionResultOk(
         .toIntArray()
 )
 
+fun ExpansionResultOk.shiftRanges(offset: Int): ExpansionResultOk {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/macros/MacroExpansionSharedCache.kt:353]:
```suggestion
fun ExpansionResultOk.shiftDstRangesRight(offset: Int): ExpansionResultOk {
```

File: src/main/kotlin/org/rust/lang/core/macros/MacroExpansionSharedCache.kt [hunk 2]
```diff
@@ -350,6 +350,12 @@ class ExpansionResultOk(
...
+    val dollarCrateOccurrences = dollarCrateOccurrences.map { it + offset }.toIntArray()
+    val ranges = ranges.ranges.map { it.copy(dstOffset = it.dstOffset + offset) }
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/macros/MacroExpansionSharedCache.kt:355]:
```suggestion
    val ranges = ranges.ranges.map { it.dstShiftRight(offset) }
```

File: src/main/kotlin/org/rust/lang/core/resolve2/DefCollector.kt [hunk 3]
```diff
@@ -373,12 +376,36 @@ class DefCollector(
...
+    private fun expandMacroInBlock(
+        callData: RsMacroCallData,
+        defData: RsMacroData,
+        mixHash: HashCode
+    ): Pair<StubElement<RsBlock>, ExpansionResultOk>? {
+        val expansion = macroExpanderShared.cachedExpand(macroExpander, defData, callData, mixHash).ok()
+            ?: return null
+        val macroExpansionContext = MacroExpansionContext.STMT
+        val expandedText = macroExpansionContext.prepareExpandedTextForParsing(expansion.text)
+        val expandedFile = RsPsiFactory(project).createPsiFile(expandedText) as? RsFile ?: return null
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/resolve2/DefCollector.kt:403]:
```suggestion
        val expandedFile = RsPsiFactory(project, markGenerated = false).createPsiFile(expandedText) as? RsFile ?: return null
```

**Review by vlad20012 — APPROVED** [review]

**Comment by dima74** [comment]
bors r=vlad20012

**Pull request merged**